Amplify/Categories/DataStore/Model/Collection/List+Model.swift

```diff
@@ -6,82 +6,97 @@
+/// `List<ModelType>` is a custom `Collection` that is capable of loading
+/// records from a data source. This is specially useful when dealing with
 /// Model associations that need to be lazy loaded.
-///
-/// When using `DataStore.query(_ modelType:)` some models might contain associations
-/// with other models and those aren't fetched automatically. This collection keeps track
-/// of the associated `id` and `field` and fetches the associated data on demand.
 public class List<ModelType: Model>: Collection, Codable, ExpressibleByArrayLiteral {
-
     public typealias Index = Int
     public typealias Element = ModelType
     public typealias Elements = [Element]
 
-    public typealias ArrayLiteralElement = ModelType
-
-    /// The array of `Element` that backs the custom collection implementation.
-    internal var elements: Elements
-
-    /// If the list represents an association between two models, the `associatedId` will
-    /// hold the information necessary to query the associated elements (e.g. comments of a post)
-    internal var associatedId: Model.Identifier?
+    /// Represents the data state of the `List`.
+    enum LoadedState {
+        case notLoaded
+        case loaded([Element])
+    }
 
-    /// The associatedField represents the field to which the owner of the `List` is linked to.
-    /// For example, if `Post.comments` is associated with `Comment.post` the `List<Comment>`
-    /// of `Post` will have a reference to the `post` field in `Comment`.
-    internal var associatedField: ModelField?
+    /// The current state of lazily loaded list
+    var loadedState: LoadedState
 
-    internal var limit: Int = 100
+    /// The provider for fulfilling list behaviors
+    let listProvider: AnyModelListProvider<Element>
 
-    /// The current state of lazily loaded list
-    internal var state: LoadState = .pending
+    /// The array of `Element` that backs the custom collection implementation.
+    /// Attempting to access the list object will attempt to retrieve the elements in memory or retrieve it from the
+    /// provider's data source. This is not thread safe as it can be performed from multiple threads, however the
+    /// provider's call to `load` should be idempotent and should result in the final loaded state.
+    var elements: [Element] {
+        get {
+            switch loadedState {
+            case .loaded(let elements):
+                return elements
+            case .notLoaded:
+                let result = listProvider.load()
+                switch result {
+                case .success(let elements):
+                    loadedState = .loaded(elements)
+                    return elements
+                case .failure(let error):
+                    Amplify.log.error(error: error)
+                    return []
+                }
+            }
+        }
+        set {
+            switch loadedState {
+            case .loaded:
+                return
```

Comment:
Log an error here also since we explicitly disallow setting list elements after they're already loaded

Reply:
And add a note on the main type doc comment